### PR TITLE
Improve profiles

### LINKED ISSUE #185: Improve CN/SAN profile language further
As [posted](https://github.com/letsencrypt/cp-cps/commit/e8bc4b86f3bd439e5663cdeaf64f72b8829c6ee6#commitcomment-107005030) by @wthayer:
> I think this language is a bit squishy. I'd suggest "May be empty or CN=one of the values from the Subject Alternative Name extension". You may also want to consider noting that the SAN MUST be marked critical when the CN is omitted.

### LINKED ISSUE #196: Clarify Issuer Distinguished Name value in Certificate Profiles
The Mozilla CA self assessment 6.1.7 Key usage purposes (as per X.509 v3 key usage field) states the following. When looking at our [CPS](https://letsencrypt.org/documents/isrg-cp-cps-v5.1/#7.1-certificate-profile), it's not explicitly clear which certificate profile would be used when determining the `Issuer Distinguished Name: Derived from Issuer certificate`. This could be made explicit and clear up some confusion that happened during the self assessment.

> Private Keys corresponding to Root Certificates MUST NOT be used to sign Certificates except in the following cases:
>
> 1. Self-signed Certificates to represent the Root CA itself;
>
> 2. Certificates for Subordinate CAs and Cross-Certified Subordinate CA Certificates;
>
> 3. Certificates for infrastructure purposes (administrative role certificates, internal CA operational device certificates); and
>
> 4. Certificates for OCSP Response verification.

An example would be, for a Subordinate CA Certificate, the `Issuer Dinstinguished Name`  data could be changed to `Derived from the Root CA Issuer certificate`.

### LINKED ISSUE #212: Miscellaneous improvements to profiles
Section 7.1:
- Subordinate: remove "Let's Encrypt Authority X<n>", as those intermediates have expired
- DV-SSL: specify that we only accept 3 specific RSA key sizes
- OCSP: remove this profile entirely, as our only delegated OCSP signer has long since expired

Section 7.2:
- Say that the signature algorithm is determined by the issuer
- Add "onlyContainsUserCerts" to the IDP in the latter of the two profiles

### LINKED ISSUE #213: Consider moving 6.1.5 Key Sizes directly into 7.1 Certificate Profiles as "Subject Public Key"
The key types and key sizes seem in [section 6.1.5](https://letsencrypt.org/documents/isrg-cp-cps-v5.2/#6.1.5-key-sizes) seem, to me, to be better suited in [section 7.1](https://letsencrypt.org/documents/isrg-cp-cps-v5.2/#7.1-certificate-profile) certificate profile in the subject public key field.

### LINKED ISSUE #217: Subordinate CA Certificate Subject Distinguished Name is overspecified
Currently, under Subordinate CA Certificate, we require that the Subject Distinguished Name has values:
C=US, O=Let's Encrypt, CN=Let's Encrypt Authority X<n>; or
C=US, O=Let's Encrypt, CN=[ER]<n>
where n is an integer representing the instance of the Subordinate CA Certificate.

While this accurately models our current names, it's a bit over-constrained: We may want to use different letter designators in the future, for example.

We should change this to allow any unique identifier per subordinate CA certificate

### LINKED ISSUE #218: Allow ECDSA P-521for subscriber certificates
It is the only ECDSA curve which the BRs allow but our CPS does not. There's no reason to limit ourselves in this way.

### PR DESCRIPTION
**Pull request by aarongable**
- Remove OCSP Delegated Responder profile, as we no longer issue such certificates
- Remove restrictions on the Common Names we set
- Remove restriction on ECDSA P-521
- Improve descriptions of key sizes and checking routines
- Improve descriptions of algorithm identifiers
- Miscellaneous formatting and phrasing improvements

Fixes #185
Fixes #196
Fixes #212
Fixes #213
Fixes #217
Fixes #218

### CODE REVIEW

File: CP-CPS.md [hunk 1]
```diff
@@ -1013,69 +1013,53 @@ See Section 5.5.5.
...
-| Extended Key Usage             | TLS Server Authentication, TLS Client Authentication                               |
-| Certificate Policies           | CAB Forum Domain Validated (2.23.140.1.2.1)<br>Optionally, ISRG Domain Validated (1.3.6.1.4.1.44947.1.1.1) with qualifier id-qt-cps pointing to this CP/CPS |
-| Authority Information Access   | Contains CA Issuers URL (and optionally an OCSP URL). URLs vary based on Issuer.   |
-| CRL Distribution Points        | Contains a CRL URL. URL varies based on Issuer.                                    |
+| Field or extension             | Value                                                                         |
+| ------------------------------ | ----------------------------------------------------------------------------- |
+| Serial Number                  | Unique, with 64 bits of output from a CSPRNG                                  |
+| Issuer Distinguished Name      | Derived from Issuer certificate                                               |
+| Subject Distinguished Name     | C=US, O=Let's Encrypt, and a meainingful CN                                   |
```
Comment (jprenken) [inline CP-CPS.md:1035]: typo for "meaningful"

**Comment by aarongable** [comment]
Let's hold on this one for a sec, I have a further improvement in the works.

File: CP-CPS.md [hunk 2]
```diff
@@ -866,22 +866,19 @@ ISRG Public Keys are also available in the Certificate Repository.
...
 ### 6.1.6 Public key parameters generation and quality checking
 
 ISRG uses HSMs conforming to FIPS 186-4, capable of providing random number generation and on-board creation of at least 2048-bit RSA keys and at least 384-bit ECDSA keys.
 
-Per NIST SP 800‐89 (<https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-89.pdf>), section 5.3.3, the CA ensures that:
+Per [NIST SP 800‐89](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-89.pdf), section 5.3.3, the CA ensures that all RSA keys in ISRG CA and Subscriber certificates have a public exponent of 65537 and an odd modulus which has no factors smaller than 752.
 
-* the public exponent of any RSA key used in a DV-SSL Certificate is in the range between 2<sup>16</sup>+1 and 2<sup>256</sup>-1, and
-* the modulus of such a certificate is an odd number, not the power of a prime, and has no factors smaller than 752.
+Per [NIST SP 800-56A (Revision 2)](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar2.pdf), Section 5.6.2.3.2, the CA ensures that all ECDSA keys in ISRG CA and Subscriber certificates comply with the ECC Full Public Key Validation Routine.
```
[the remark this answers is not in the document]
Comment (aarongable) [reply CP-CPS.md:881]: Good to know, but the BRs cite rev 2.

File: CP-CPS.md [hunk 3]
```diff
@@ -866,22 +866,19 @@ ISRG Public Keys are also available in the Certificate Repository.
 
 ### 6.1.5 Key sizes
 
-ISRG Root CA RSA Private Keys are at least 4096 bits in length.
+ISRG Root CA keys are either RSA with a modulus of length 4096 or ECDSA with namedCurve P-384.
```
Comment (jsha) [inline CP-CPS.md:869]: This omits "bits" so length is ambiguous. Bytes?

Probably should mirror the BRs:

https://cabforum.org/working-groups/server/baseline-requirements/documents/CA-Browser-Forum-TLS-BR-2.0.7.pdf

> 6.1.5 Key sizes
> For RSA key pairs the CA SHALL:
> • Ensure that the modulus size, when encoded, is at least 2048 bits, and;
> • Ensure that the modulus size, in bits, is evenly divisible by 8.
> For ECDSA key pairs, the CA SHALL:
> • Ensure that the key represents a valid point on the NIST P‐256, NIST P‐384 or NIST P‐521
elliptic curve.

File: CP-CPS.md [hunk 4]
```diff
@@ -1089,14 +1072,23 @@ See section 7.1.
...
+| ECDSA P-384 | `301006072a8648ce3d020106052b81040022`       |
+| ECDSA P-521 | `301006072a8648ce3d020106052b81040023`       |
 
 #### 7.1.3.2 Signature AlgorithmIdentifier
 
-| Name                    | Object identifier                    |
-| ----------------------- | ------------------------------------ |
-| sha256WithRSAEncryption | 1.2.840.113549.1.1.11                |
-| ecdsa-with-SHA384       | 1.2.840.10045.4.3.3                  |
+The `AlgorithmIdentifier` field of `signature` and `signatureAlgorithm` fields of ISRG Certificates, CRLs, and OCSP Responses has the following (hex-encoded) DER bytes, corresponding to the issuer's key type and signature algorithm:
```
Comment (jsha) [inline CP-CPS.md:1086]: BRs 7.1.3.2:

> 7.1.3.2 Signature AlgorithmIdentifier
> All objects signed by a CA Private Key MUST conform to these requirements on the use of the
> AlgorithmIdentifier or AlgorithmIdentifier‐derived type in the context of signatures.
> In particular, it applies to all of the following objects and fields:
> • The signatureAlgorithm field of a Certificate or Precertificate.
> • The signature field of a TBSCertificate (for example, as used by either a Certificate or
Precertificate).
> • The signatureAlgorithm field of a CertificateList
> • The signature field of a TBSCertList
> • The signatureAlgorithm field of a BasicOCSPResponse.

The language in this draft misses some objects / fields. We should mirror the BRs:

```suggestion
The encoding of the `AlgorithmIdentifier` field of `signature` and `signatureAlgorithm` fields of the following objects and fields uses the hex-encoded DER bytes specified below, corresponding to key type:

 • The signatureAlgorithm field of a Certificate or Precertificate.
 • The signature field of a TBSCertificate (for example, as used by either a Certificate or
Precertificate).
 • The signatureAlgorithm field of a CertificateList
 • The signature field of a TBSCertList
 • The signatureAlgorithm field of a BasicOCSPResponse.
```

Also the list of key type / algorithm identifier should include ECDSA P-256. We don't currently have any P-256 intermediates but we can specify which AlgorithmIdentifier we would use if we did have them.

**Review by jsha — CHANGES_REQUESTED** [review]
One comment in this review, and also one I previously left as an independent comment: https://github.com/letsencrypt/cp-cps/pull/228/files?w=1#r1811208144

File: CP-CPS.md (hunk 3 above)
Comment (aarongable) [reply CP-CPS.md:869]: Done. I've borrowed the phrasing from the BRs, but kept it as prose (not bulleted).

File: CP-CPS.md (hunk 4 above)
Comment (aarongable) [reply CP-CPS.md:1086]: I disagree that anything has been missed: "`signature` and `signatureAlgorithm` fields of ISRG Certificates" captures the first two bullet points as both of those fields are in Certificates, and Precertificates are a kind of Certificate. The same holds for the two bullet points representing CRLs and the one bullet point covering OCSP.

That said, listing these doesn't actually help: the five bullet points in BRs 7.1.3.2 are actually examples, and the requirement is on "All objects signed by a CA" so maybe we should just say that?

I don't think we should specify the AlgorithmIdentifier we would use for P-256 intermediates, unless we also update Section 6.1.5 to say that we have P-256 intermediates.

Comment (jsha) [reply CP-CPS.md]: Yeah, I like just saying "all objects signed by the CA."

**Review by jsha — APPROVED** [review]

**Review by beautifulentropy — APPROVED** [review]

**Review by jprenken — APPROVED** [review]

**Review by bdaehlie — APPROVED** [review]

**Pull request merged**